zarr/meta.py

```diff
@@ -71,10 +71,13 @@ def encode_array_metadata(meta):
 
 
 def encode_dtype(d):
-    if d.fields is None:
-        return d.str
-    else:
+    if d.fields is not None:
         return d.descr
+    elif d.subdtype is not None:
+        sdname, sdshape = d.subdtype
+        return "%s%s" % (str(sdshape), sdname.str)
```

Comment:
Should this remain a void type à la `V16`, or maintain the provided shape and type?

Reply:
"In numpy, if you make an unstructured array with a dtype with a shape, the shape from the dtype gets pushed into the shape of the array."